proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2521,6 +2439,445 @@ lemma possibleSwitchTo_all_invs_but_ct_not_inQ':
   apply (intro conjI; fastforce?)
   by (metis fold_list_refs_of_replies')
 
+lemma possibleSwitchTo_sch_act_not_other:
+  "\<lbrace>\<lambda>s. sch_act_not t' s \<and> t' \<noteq> t\<rbrace>
+   possibleSwitchTo t
+   \<lbrace>\<lambda>_. sch_act_not t'\<rbrace>"
+  apply (clarsimp simp: possibleSwitchTo_def)
+  apply (wpsimp wp: threadGet_wp inReleaseQueue_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+(* FIXME RT Michael: also in Victor's PR *)
+lemma replyUnlink_weak_sch_act_wf[wp]:
+  "\<lbrace>\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s \<and> sch_act_not t s\<rbrace>
+   replyUnlink r t
+   \<lbrace>\<lambda>_ s. weak_sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  unfolding replyUnlink_def setReplyTCB_def getReplyTCB_def
+  by (wpsimp wp: hoare_vcg_imp_lift hoare_vcg_all_lift gts_wp'
+           simp: weak_sch_act_wf_def)
+
+lemma restart_thread_if_no_fault_reply_tcb_reply_at[wp]:
+  "restart_thread_if_no_fault t  \<lbrace>reply_tcb_reply_at ((=) (Some t')) r_opt\<rbrace>"
+  apply (clarsimp simp: sk_obj_at_pred_def)
+  apply (wpsimp wp: restart_thread_if_no_fault_obj_at_impossible')
+  done
+
+lemma reply_unlink_tcb_reply_tcb_reply_at_other:
+  "\<lbrace>\<lambda>s. reply_tcb_reply_at ((=) (Some t')) r' s \<and> t' \<noteq> t\<rbrace>
+   reply_unlink_tcb t r
+   \<lbrace>\<lambda>_ s. reply_tcb_reply_at ((=) (Some t')) r' s\<rbrace>"
+  apply (wpsimp wp: update_sk_obj_ref_wps get_simple_ko_wp gts_wp
+              simp: reply_unlink_tcb_def)
+  apply (clarsimp simp: reply_tcb_reply_at_def obj_at_def pred_tcb_at_def)
+  done
+
+abbreviation (input)
+  "reply_unlink_ts_pred t s
+    \<equiv> \<forall>ep r_opt pl.
+       pred_map ((=) (Structures_A.thread_state.BlockedOnReceive ep (Some r_opt) pl)) (tcb_sts_of s) t
+       \<longrightarrow> reply_tcb_reply_at ((=) (Some t)) r_opt s"
+
+abbreviation (input)
+  "cancel_all_ipc_loop_body t
+    \<equiv> do st \<leftarrow> get_thread_state t;
+          reply_opt \<leftarrow> case st of Structures_A.thread_state.BlockedOnReceive x r_opt xa
+                                    \<Rightarrow> return r_opt
+                                | _ \<Rightarrow> return None;
+          y \<leftarrow> when (reply_opt \<noteq> None) $ reply_unlink_tcb t (the reply_opt);
+          restart_thread_if_no_fault t
+       od"
```

Comment:
In an out-of-band chat, @michaelmcinerney commented that getting helper lemmas to apply can sometimes be fiddly. For example, in the proof of `cancel_all_ipc_corres`, we apply `cancel_all_ipc_corres_helper`. The latter is phrased in terms of this and the following abbreviation. This means that the abbreviations need to be phrased to exactly match the whatever the goal happens to be at the point the helper lemma is applied. The abbreviation is therefore sensitive to the preceding proof steps.

An alternative is to make this a `definition`. This requires extra steps to unfold the definition in the helper lemma proof, and to fold it in the `corres` proof. On the other hand, it allows you to do that folding much earlier than the point where you apply the helper lemma. Then the body of the definition is not longer sensitive to the intervening proof steps. In particular, you can fold before you do any proof steps. Then `cancel_all_ipc_loop_body` can look exactly like the corresponding part of the spec.

If you want, you can then take this a step further, and move `cancel_all_ipc_loop_body` all the way up to the spec, and use it in the definition of `cancel_all_ipc`. That would require you to update some other proofs, but there are ways around that, too, e.g.
```
definition cancel_all_ipc where
  cancel_all_ipc_def2: "cancel_all_ipc epptr == ..."

lemmas cancel_all_ipc_def = cancel_all_ipc_def2[unfolded cancel_all_ipc_loop_body_def]
```

Reply:
Very good points. The thought of making this a definition did cross my mind (like the aborted `valid_dom_schedule'` experiment) but my mind immediately jumped to putting this in the spec, and I wasn't so sure about that. I think I'll go for making them definitions in Refine (with a FIXME RT: for the abstract version), but I'll mull it over a bit more